### PR TITLE
Write backwards compatible row group statistics (#3526)

### LINKED ISSUE #799: parquet min_max statistics cannot read by `pyarrow`
**Describe the bug**
a parquet file created by `arrow-rs` has no min_max statistics when reading by `pyarrow`.

**To Reproduce**
```rust
// rust code

let id_array = Int32Array::from(vec![1, 2, 3, 4, 5]);
let id_array2 = Int32Array::from(vec![2, 3, 4, 5, 6]);
let schema = Arc::new(Schema::new(vec![
    Field::new("id", DataType::Int32, false),
    Field::new("id2", DataType::Int32, false),
]));

let batch = RecordBatch::try_new(
    schema.clone(),
    vec![Arc::new(id_array), Arc::new(id_array2)],
)
.unwrap();

let writer_properties = WriterProperties::builder()
    .set_compression(Compression::ZSTD)
    .set_statistics_enabled(true)
    .build();

let path = "/.../test.parquet";
let file = fs::File::create(&path).unwrap();

let mut writer = ArrowWriter::try_new(file, schema.clone(), Some(writer_properties)).unwrap();
writer.write(&batch).unwrap();
writer.close().unwrap();

let file2 = fs::File::open(&path).unwrap();

let file_reader = SerializedFileReader::new(file2).unwrap();
let mut arrow_reader = ParquetFileArrowReader::new(Arc::new(file_reader));

println!(
    "statistics: {:?}",
    arrow_reader
        .get_metadata()
        .row_group(0)
        .column(0)
        .statistics()
);
println!(
    "statistics: {:?}",
    arrow_reader
        .get_metadata()
        .row_group(0)
        .column(1)
        .statistics()
);

// output: 
// statistics: Some(Int32({min: Some(1), max: Some(5), distinct_count: None, null_count: 0, min_max_deprecated: false}))
// statistics: Some(Int32({min: Some(2), max: Some(6), distinct_count: None, null_count: 0, min_max_deprecated: false}))
```

```python
# python code

import pyarrow.parquet as pq
f = pq.ParquetFile('./test.parquet')
print(f.metadata.row_group(0).column(0).statistics)

# output:
"""
<pyarrow._parquet.Statistics object at 0x7fbf8d409dd0>
  has_min_max: False
  min: None
  max: None
  null_count: 0
  distinct_count: 0
  num_values: 5
  physical_type: INT32
  logical_type: None
  converted_type (legacy): NONE
"""
```

**Expected behavior**
pyarrow should get statistics like
```
  has_min_max: True
  min: 1
  max: 5
```

**Additional context**
rust lib version:
```
parquet = "5.4.0"
arrow = "5.4.0"
```

python lib version:
```
pyarrow==5.0.0
```

### LINKED ISSUE #3526: Write Deprecated Min Max Statistics When ColumnOrder Signed
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

As per [the specification](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L200) for backwards compatibility it is suggested to write the old min/max statistics.

**Describe the solution you'd like**

If the column order is signed and the writer version is v1, we should write both statistics variants.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #3526
Closes #799

# Rationale for this change
 

This helps with compatibility with pyarrow, which doesn't understand min_value yet https://github.com/apache/arrow/issues/13976

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
Test failures fixed in #3528

**Comment by alamb** [comment]
I think this may also fix https://github.com/apache/arrow-rs/issues/799 filed by @tfiasco and which @pjmore  also commented on

File: parquet/src/file/statistics.rs [hunk 1]
```diff
@@ -329,6 +331,12 @@ impl Statistics {
         statistics_enum_func![self, is_min_max_deprecated]
     }
 
+    /// Returns `true` if the statistics are backwards compatible with the
+    /// deprecated `min` and `max` fields
```
Comment (alamb) [inline parquet/src/file/statistics.rs:335]:
```suggestion
    /// deprecated `min` and `max` fields in parquet version `1.0` spec
```

File: parquet/src/column/writer/mod.rs [hunk 2]
```diff
@@ -817,13 +817,16 @@ impl<'a, E: ColumnValueEncoder> GenericColumnWriter<'a, E> {
...
         if self.statistics_enabled != EnabledStatistics::None {
-            let statistics = Statistics::new(
+            let statistics = ValueStatistics::<E::T>::new(
                 self.column_metrics.min_column_value.clone(),
                 self.column_metrics.max_column_value.clone(),
                 self.column_metrics.column_distinct_count,
                 self.column_metrics.num_column_nulls,
                 false,
             );
+            let statistics = statistics
```
Comment (alamb) [inline parquet/src/column/writer/mod.rs:827]:
```suggestion
            // Some common readers only support "old" statistics format
            // so write them out if possibe. See 
            // https://github.com/apache/arrow-rs/issues/799
            let statistics = statistics
```

File: parquet/src/column/writer/mod.rs [hunk 3]
```diff
@@ -1893,6 +1896,7 @@ mod tests {
     fn test_bool_statistics() {
         let stats = statistics_roundtrip::<BoolType>(&[true, false, false, true]);
         assert!(stats.has_min_max_set());
+        assert!(!stats.is_min_max_backwards_compatible());
```
Comment (alamb) [inline parquet/src/column/writer/mod.rs:1899]: is it just that bool is not backwards compatibel?

File: parquet/src/file/statistics.rs [hunk 4]
```diff
@@ -252,10 +252,12 @@ pub fn to_thrift(stats: Option<&Statistics>) -> Option<TStatistics> {
         (None, None)
     };
 
-    if stats.is_min_max_deprecated() {
-        thrift_stats.min = min;
-        thrift_stats.max = max;
-    } else {
+    if stats.is_min_max_backwards_compatible() {
```
Comment (alamb) [inline parquet/src/file/statistics.rs:255]:
```suggestion
    // Copy to min/max values from parquet 1.0 for compatibility
    // with other writers
    if stats.is_min_max_backwards_compatible() {
```

File: parquet/src/file/statistics.rs [hunk 5]
```diff
@@ -405,7 +413,13 @@ pub struct ValueStatistics<T> {
     // Distinct count could be omitted in some cases
     distinct_count: Option<u64>,
     null_count: u64,
+
+    /// If `true` populate the deprecated `min` and `max` fields instead of
+    /// `min_value` and `max_value`
     is_min_max_deprecated: bool,
+
+    /// If `true` should always write deprecated `min` and `max` fields
```
Comment (alamb) [inline parquet/src/file/statistics.rs:421]: is this in addition to `min_value` and `max_value`?

I found it hard to understand how these two flags were supposed to interact

File: parquet/src/file/statistics.rs [hunk 6]
```diff
@@ -423,6 +437,15 @@ impl<T: ParquetValueType> ValueStatistics<T> {
             distinct_count,
             null_count,
             is_min_max_deprecated,
+            is_min_max_backwards_compatible: is_min_max_deprecated,
+        }
+    }
+
+    /// Set whether to write the deprecated `min` and `max` fields
```
Comment (alamb) [inline parquet/src/file/statistics.rs:444]:
```suggestion
    /// Set whether to write the deprecated `min` and `max` fields
    /// for compatibility with older parquet writers
```

File: parquet/src/file/statistics.rs [hunk 7]
```diff
@@ -478,6 +501,12 @@ impl<T: ParquetValueType> ValueStatistics<T> {
     fn is_min_max_deprecated(&self) -> bool {
         self.is_min_max_deprecated
     }
+
+    /// Returns `true` if the statistics are backwards compatible
```
Comment (alamb) [inline parquet/src/file/statistics.rs:505]: is it that the statistics are backwards compatible or they are desired to be?

**Review by alamb — APPROVED** [review]
LGTM -- thank you @tustvold . I think this will be a very nice feature addition for othr users. 

I had a bunch of naming / doc / style suggestions but nothing required I don't think

I also tested the reproducer from https://github.com/apache/arrow-rs/issues/799 from @tfiasco and the min/max values are properly written 👍 


```
In [3]: 
   ...: import pyarrow.parquet as pq
   ...: f = pq.ParquetFile('/tmp//test.parquet')
   ...: print(f.metadata.row_group(0).column(0).statistics)
<pyarrow._parquet.Statistics object at 0x110118c70>
  has_min_max: True
  min: 1
  max: 5
  null_count: 0
  distinct_count: 0
  num_values: 5
  physical_type: INT32
  logical_type: None
  converted_type (legacy): NONE

In [4]: 
```

File: parquet/src/column/writer/mod.rs (hunk 3 above)
Comment (tustvold) [reply parquet/src/column/writer/mod.rs:1905]: Yup, added a doc to that effect

**Comment by tustvold** [comment]
I've updated the docs to hopefully be a bit clearer, PTAL.

I dropped the mentions of parquet v1, as I'm not entirely sure when the deprecation actually happened - the parquet changelog isn't particularly clear on this aspect...

**Comment by alamb** [comment]
Thanks @tustvold

**Pull request merged**